languages/ruby/reference/exercise-concepts/flatten-array.md

```diff
@@ -0,0 +1,29 @@
+# Concepts of flatten-array
+
+## General
+
+- method definition: the student must implement the `flatten` method
+- method arguments (positional): `flatten` takes one argument
+- return values: `flatten` should return an array
+- class methods: one approach is the define a class `FlattenArray` with a class method `flatten`
+- module methods: one approach is to implement `FlattenArray` as a module
+- arrays: the input is an array, and student's should understand the basics of arrays
+- array flatten: this method is the obvious/preferred way to flatten an array
```

Comment:
really :o ? I thought that the use of builtins was encouraged when applicable. Maybe this is something that should be added to the exercise instructions if it's the case?

Reply:
Normally, it is... in this case, we are building something that Ruby already has built in... so the exercise is not, perhaps, worthwhile if we use the built in method.  Kind of like not using `Date.leap?` for the leap year exercise.

Reply:
ok, so should I maybe list recursion here as a concept then? It feels like the next best contender.

I'm not sure what the *purpose* of discouraging the use of `flatten` is :/ it's a method born out of a common need, so why should we push the students to solve the problem using nonidiomatic means.

Reply:
I guess I don't know what the exercise is, then. 

If it is about deep nesting, and flattening a deeply nested array to be a single level array, then yes, by all means, use flatten for the "light work".  But if it is mean to show that we can implement behavior that we have in Ruby, and it might be an interesting way to learn the language, then we would discourage the method that we are replacing.

Reply:
I agree - it all depends on what the learning goal for this exercise is. I can be either to
1. Discover the built in `flatten` method, and add it to the students "toolbox"
2. Learn to implement their own `flatten` method

Imo both are valid goals, but we should specify the goal in the docs. Maybe if we go with 2. then we should tell the student that a built in method exists, but they should implement their own version of the method. That way they'll both get the benefit of learning about the existence of `Array#flatten` (if they didn't already know about it) and also learn to implement their own version of it.

Reply:
Check the Ruby repository and see the discussion concerning this exercise, as well as where we do limit the use of the built in for the Leap exercise.

Reply:
Thanks, I took a look at the PR where the issue was discussed, and it seems like the consensus was that the students should write their own implementation 👍 

What do you think about opening an issue about adding this information to the exercise instructions? Let the students know that the builtin exists (good for them if they didn't already know about it) but instruct them to write their own version.

Reply:
I think part of having gotten to this point is having the sophistication to have the documentation in hand.  For v2 it is a discussion point at the time of the review.

Perhaps finding methods should be a skill that is at or before the first concept exercise?

Reply:
I'm sure that most of the students would be capable of finding the method on their own. 

The purpose of giving them the heads up in this case would be to eliminate one review step. If most of the students submit a first iteration which uses the builtin `flatten`, only to be told by a mentor that the purpose is to write their own implementation, then we've only wasted the students' and mentors' time.

We could also instruct them: "(you are meant to) write your own implementation" without actually mentioning the existence of `Array#flatten`

Reply:
This is v3 though, so this is automated... not v2 where a mentor will see this and mentor it.

Unless I am missing something here.

I think it does fit to mention `Array#flatten` but for this exercise, is it the preferred way?  And is it the obvious way?  The full solution, it appears, would be `flatten.compact`.

Reply:
For v3 it definitely doesn't matter that much since the review is mostly automated as you said.

>I think it does fit to mention Array#flatten but for this exercise, is it the preferred way?

I think that we're currently in agreement that `Array#flatten` is not the intended solution for this exercise as it's a bit too trivial to do in ruby, but it could be mentioned. I'll update list 👍 

Outside of this exercise I would definitely regard it as the obvious solution :) If I saw - in the wild - someone writing their own implementation of the method, I'd definitely wonder if there was some special reason why the builtin wasn't used